### PR TITLE
feat: allow authz granters to specify forwarding info for token transfer

### PR DESCRIPTION
**Pull request by bznein**
## Description

closes: #6589

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

### CODE REVIEW

**Comment by bznein** [comment]
@crodriguezvega this draft PR targets issue #6589, I'm not sure if we want to merge this already and then do the other changes (possibly naming, and those in #6624) on top, or if we want to hold off on this one and merge it when we decided on everything.

**Comment by bznein** [comment]
Discussed it offline with Carlos and agreed on having this out for review and then modify the code as needed later.

File: modules/apps/transfer/types/transfer_authorization_test.go [hunk 1]
```diff
@@ -389,6 +389,84 @@ func (suite *TypesTestSuite) TestTransferAuthorizationAccept() {
...
+					},
+				}
+			},
+			func(res authz.AcceptResponse, err error) {
+				suite.Require().Error(err)
+				suite.Require().False(res.Accept)
+			},
+		},
+		{
+			"failure - Allocation specify hops but msgTransfer does not have hops",
```
Comment (bznein) [inline modules/apps/transfer/types/transfer_authorization_test.go:454]: I'm not sure I interpreted the requirements correctly. Is this a right case or should the `Accept` succeed if the allocation specifies allowed hops and the forwarding field in the `msgTransfer` doesn't have it?

File: modules/apps/transfer/types/transfer_authorization.go [hunk 2]
```diff
@@ -171,6 +176,21 @@ func isAllowedAddress(ctx sdk.Context, receiver string, allowedAddrs []string) b
 	return false
 }
 
+// areHopsAllowed returns whether the provided slice of Hop matches one of the allowed ones.
+func areHopsAllowed(provided []Hop, allowed []*Hops) bool {
+	if len(allowed) == 0 {
```
Comment (damiannolan) [inline modules/apps/transfer/types/transfer_authorization.go:181]: shouldn't this be checking len of provided (i.e. msgTransfer hops)? 

If the length of the msg hops is zero, then it should by default be allowed. Otherwise, we must ensure that the provided hops are specified within the authz allocation.

File: modules/apps/transfer/types/transfer_authorization.go [hunk 3]
```diff
@@ -171,6 +176,21 @@ func isAllowedAddress(ctx sdk.Context, receiver string, allowedAddrs []string) b
 	return false
 }
 
+// areHopsAllowed returns whether the provided slice of Hop matches one of the allowed ones.
+func areHopsAllowed(provided []Hop, allowed []*Hops) bool {
```
Comment (damiannolan) [inline modules/apps/transfer/types/transfer_authorization.go:180]: nit, maybe would go for something like `isAllowedForwarding`.. food for thought. Happy to hear input from others as well!

File: modules/core/errors/errors.go [hunk 4]
```diff
@@ -60,4 +60,7 @@ var (
 
 	// ErrNotFound defines an error when requested entity doesn't exist in the state.
 	ErrNotFound = errorsmod.Register(codespace, 16, "not found")
+
+	// ErrUnauthorizedHops defines an error when the hops provided are not authorized.
+	ErrUnauthorizedHops = errorsmod.Register(codespace, 17, "unauthorized hops")
```
Comment (damiannolan) [inline modules/core/errors/errors.go:65]: Would suggest to maintain this error within `transfer/types` pkg

File: proto/ibc/applications/transfer/v1/authz.proto [hunk 5]
```diff
@@ -22,6 +23,14 @@ message Allocation {
   // allow list of memo strings, an empty list prohibits all memo strings;
   // a list only with "*" permits any memo string
   repeated string allowed_packet_data = 5;
+  // a list of Allowed hops
+  repeated Hops allowed_forwarding_hops = 6;
+}
+
+
+// Hops represent a list of Hop
+message Hops {
```
Comment (damiannolan) [inline proto/ibc/applications/transfer/v1/authz.proto:32]: Is there any specific reason to introduce a new type here wrapping `Hop`? couldn't we use a `repeated Hop` field within the wrapper inside of `Allocation`?

File: modules/apps/transfer/types/transfer_authorization.go (hunk 3 above)
Comment (bznein) [reply modules/apps/transfer/types/transfer_authorization.go:180]: I think that makes sense, more in line with the rest on the function

File: modules/core/errors/errors.go (hunk 4 above)
Comment (bznein) [reply modules/core/errors/errors.go:65]: Done

File: proto/ibc/applications/transfer/v1/authz.proto (hunk 5 above)
Comment (bznein) [reply proto/ibc/applications/transfer/v1/authz.proto:32]: I'm not sure I follow: `allowed_forwarding_hops` is a list of list of Hops. So we can't replace it with a `repeated Hop`. It is a `repeated repeated Hop`.

But let me know if I didn't get your comment!

File: modules/apps/transfer/types/transfer_authorization.go (hunk 2 above)
Comment (bznein) [reply modules/apps/transfer/types/transfer_authorization.go:181]: Two things here

* Since `allowed_forwarding_hops` is not a mandatory field, we check whether it's empty because, in that case, all hops are allowed (in my understanding

* > If the length of the msg hops is zero, then it should by default be allowed.
That's the part I'm not sure and I asked the same here: https://github.com/cosmos/ibc-go/pull/6627/files#r1645684477 
If `allowed_forwarding_hops` is not empty, but the msg contains no hops, should we accept it?

Comment (gjermundgaraba) [reply modules/apps/transfer/types/transfer_authorization.go:181]: I would like a definitive answer on this as well, but my understanding here was that if the allowed hops are set, then those hops needs to be part of the message. Meaning, you shouldn't be allowed to send through any other paths (including one without hops). To me, the allowed_forwarding_hops guarantees that when you send a token, it will end up on the expected chain with the expected denom. If that is not the case, it would be good to get that clarified :)

File: modules/apps/transfer/types/transfer_authorization_test.go [hunk 6]
```diff
@@ -389,6 +389,84 @@ func (suite *TypesTestSuite) TestTransferAuthorizationAccept() {
 				suite.Require().Nil(res.Updated)
 			},
 		},
+		{
+			"success - allowed forwarding hops",
+			func() {
+				msgTransfer.Forwarding = types.NewForwarding("", types.Hop{PortId: "1", ChannelId: "channel-1"}, types.Hop{PortId: "2", ChannelId: "channel-2"})
```
Comment (gjermundgaraba) [inline modules/apps/transfer/types/transfer_authorization_test.go:395]: nit: would it make sense to use the same port as the other tests for consistency `ibctesting.MockPort`?

File: modules/apps/transfer/types/transfer_authorization_test.go [hunk 7]
```diff
@@ -389,6 +389,84 @@ func (suite *TypesTestSuite) TestTransferAuthorizationAccept() {
...
+					},
+				}
+			},
+			func(res authz.AcceptResponse, err error) {
+				suite.Require().NoError(err)
+				suite.Require().True(res.Accept)
+			},
+		},
+		{
+			"failure - allowed forwarding hops is longer",
```
Comment (crodriguezvega) [inline modules/apps/transfer/types/transfer_authorization_test.go:411]: Maybe something like this?

```suggestion
			"failure - allowed forwarding hops contains more hops",
```

File: modules/apps/transfer/types/transfer_authorization_test.go [hunk 8]
```diff
@@ -389,6 +389,84 @@ func (suite *TypesTestSuite) TestTransferAuthorizationAccept() {
...
+					},
+				}
+			},
+			func(res authz.AcceptResponse, err error) {
+				suite.Require().Error(err)
+				suite.Require().False(res.Accept)
+			},
+		},
+		{
+			"failure - One of the hops is different",
```
Comment (crodriguezvega) [inline modules/apps/transfer/types/transfer_authorization_test.go:433]:
```suggestion
			"failure - allowed forwarding hops contains one different hop",
```

File: modules/apps/transfer/types/transfer_authorization_test.go [hunk 9]
```diff
@@ -389,6 +389,84 @@ func (suite *TypesTestSuite) TestTransferAuthorizationAccept() {
...
+							{PortId: "2", ChannelId: "channel-2"},
+						},
+					},
+				}
+			},
+			func(res authz.AcceptResponse, err error) {
+				suite.Require().Error(err)
+				suite.Require().False(res.Accept)
+			},
+		},
```
Comment (crodriguezvega) [inline modules/apps/transfer/types/transfer_authorization_test.go:469]: We could also add an extra failure test where `allowed_forwarding_hops` is empty and `hops` in `MsgTransfer` is not empty.

**Review by crodriguezvega — COMMENTED** [review]
Thank you, @bznein! Sorry if the requirements were not clear, I hope I have answered your questions and it is now clearer.

File: modules/apps/transfer/types/transfer_authorization.go [hunk 10]
```diff
@@ -52,6 +53,10 @@ func (a TransferAuthorization) Accept(goCtx context.Context, msg proto.Message)
 		return authz.AcceptResponse{}, errorsmod.Wrapf(ibcerrors.ErrNotFound, "requested port and channel allocation does not exist")
 	}
 
+	if !isAllowedForwarding(msgTransfer.Forwarding.Hops, a.Allocations[index].AllowedForwardingHops) {
+		return authz.AcceptResponse{}, errorsmod.Wrapf(ErrUnauthorizedHops, "forwarding hops are not authorized")
```
[the remark this answers is not in the document]
Comment (bznein) [reply modules/apps/transfer/types/transfer_authorization.go]: Good idea, done!

File: modules/apps/transfer/types/transfer_authorization.go [hunk 11]
```diff
@@ -171,6 +176,21 @@ func isAllowedAddress(ctx sdk.Context, receiver string, allowedAddrs []string) b
 	return false
 }
 
+// isAllowedForwarding returns whether the provided slice of Hop matches one of the allowed ones.
+func isAllowedForwarding(provided []Hop, allowed []*Hops) bool {
```
[the remark this answers is not in the document]
Comment (bznein) [reply modules/apps/transfer/types/transfer_authorization.go]: Done!

File: modules/apps/transfer/types/transfer_authorization.go (hunk 2 above)
Comment (bznein) [reply modules/apps/transfer/types/transfer_authorization.go]: Done (I didn't add the `if len(allowed) == 0`) because that's implicit (if `provided` is not empty, but `allowed` is, we directly skip the for loop and return false. 

I've also added a test case to cover it.

Thanks for the clarification!

File: modules/apps/transfer/types/transfer_authorization_test.go [hunk 12]
```diff
@@ -389,6 +389,84 @@ func (suite *TypesTestSuite) TestTransferAuthorizationAccept() {
 				suite.Require().Nil(res.Updated)
 			},
 		},
+		{
+			"success - allowed forwarding hops",
```
[the remark this answers is not in the document]
Comment (bznein) [reply modules/apps/transfer/types/transfer_authorization_test.go]: Done!

File: modules/apps/transfer/types/transfer_authorization_test.go (hunk 6 above)
Comment (bznein) [reply modules/apps/transfer/types/transfer_authorization_test.go]: Yeah good idea, done :)

File: modules/apps/transfer/types/transfer_authorization_test.go (hunk 7 above)
Comment (bznein) [reply modules/apps/transfer/types/transfer_authorization_test.go]: Done!

File: modules/apps/transfer/types/transfer_authorization_test.go (hunk 8 above)
Comment (bznein) [reply modules/apps/transfer/types/transfer_authorization_test.go]: Done!

File: modules/apps/transfer/types/transfer_authorization_test.go (hunk 9 above)
Comment (bznein) [reply modules/apps/transfer/types/transfer_authorization_test.go:488]: Added!

File: modules/apps/transfer/types/transfer_authorization.go (hunk 2 above)
Comment (gjermundgaraba) [reply modules/apps/transfer/types/transfer_authorization.go]: The only potential risk there is that would allow someone to spoof the address (e.g. a fake chain) and send it there. The only way to truly enforce that you can only send to the correct chain would then be to use both port/channel (to avoid the case above) and hops, right? 

I can see both sides of this, and neither is 100% intuitive (to me). So I think we just need to carefully document the behavior so nobody gets surprised :)

**Comment by bznein** [comment]
I made a git mess :) pls ignore this PR until I fix it :)
